spring-xd-dirt/src/main/java/org/springframework/xd/dirt/plugins/spark/streaming/MessageBusReceiver.java

```diff
@@ -121,7 +165,8 @@ public MessageStoringChannel() {
 
 		@Override
 		protected boolean doSend(Message<?> message, long timeout) {
-			store(message.getPayload());
+			blockGenerator.addDataWithCallback(message.getPayload(), message.getHeaders());
+			headersList.add(message.getHeaders());
```

Comment:
This looks like a race condition - we could get the callback before we add the headers.

Even if you swap the statements, consider concurrency; it would be possible that one thread puts its header in first, while the second initiates the push.

You might have to synchronize this entire method.